### PR TITLE
Story2

### PR DESCRIPTION
**Pull request by Henry-Tottle**
Before trainer review, do peer review

### CODE REVIEW

File: categories.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
+<?php
+require_once ('vendor/autoload.php');
+header('Content-Type: application/json; charset=utf-8');
+header("Access-Control-Allow-Origin: *");
```
Comment (matt-ioa) [inline categories.php]: I think it would be worth creating a header service that returns these, to make things a bit more reusable

File: categories.php [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+use \FurnitureStoreApi\Categories\CategoryHydrator;
+
+try
+{
+    $results = ['message'=>'Successfully retrieved products',
+        'data' => CategoryHydrator::getCategories()];
+    echo json_encode($results);
+    http_response_code(200);
```
Comment (matt-ioa) [inline categories.php]: Putting the results into `['message'=>, 'data'=>]` format is something that will happen quite often, so it's worth creating a `ResponseService` with a static method to take a message string and json string, and create a response from them.

Comment (matt-ioa) [reply categories.php]: It should also take a number for the response code and set that.

File: furnitureStoreDBPopulate.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+//RUN ONCE AFTER DB (furniture_store) STRUCTURE HAS BEEN BUILT
+require('vendor/autoload.php');
+
+$jsonRequest = file_get_contents('https://dev.io-academy.uk/resources/furniture/furniture.json');
```
Comment (matt-ioa) [inline furnitureStoreDBPopulate.php]: To make your code more self-contained, it would be better to download the json file and host it locally: that way the database could be recreated even if our servers go down!

File: furnitureStoreDBPopulate.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+$json = json_decode($jsonRequest, true);
+try
+{
+    $db = new PDO('mysql:host=db;dbname=furniture_store', 'root', 'password');
```
Comment (matt-ioa) [inline furnitureStoreDBPopulate.php]: The rest of the code could also be inside the `try` block, just in case

File: furnitureStoreDBPopulate.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+}
+catch (Exception $exception)
+{
+    $logPath = 'error.log';
+    $errorMessage = PHP_EOL.$exception->getMessage();
+    file_put_contents($logPath,$errorMessage, FILE_APPEND);
+    echo '<h1>Database Connection Error, look at error.log</h1>';
+}
+
```
Comment (matt-ioa) [inline furnitureStoreDBPopulate.php]: I highly recommend adding code to add the category IDs into the database as well.

File: src/Categories/CategoryEntity.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
+<?php
+namespace FurnitureStoreApi\Categories;
+header('Content-Type: application/json; charset=utf-8');
```
Comment (matt-ioa) [inline src/Categories/CategoryEntity.php]: This header should be set in your `categories.php`, ideally using a service (as mentioned before)

File: src/Categories/CategoryEntity.php [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        return $this->id;
+    }
+
+    public function getCategory(): string
+    {
+        return $this->category;
+    }
+
+    public function getCategoryid(): int
```
Comment (matt-ioa) [inline src/Categories/CategoryEntity.php]: Is this really returning the categoryID, or is it actually the count of products in that category?

File: src/Categories/CategoryHydrator.php [hunk 8]
```diff
@@ -0,0 +1,18 @@
+<?php
+namespace FurnitureStoreApi\Categories;
+use \PDO;
+
+class CategoryHydrator
+{
+    public static function getCategories()
+    {
+        $db = new PDO('mysql:host=db;dbname=furniture_store', 'root', 'password');
```
Comment (matt-ioa) [inline src/Categories/CategoryHydrator.php]: Again, you don't want to create a new database connection every time this method runs! Either create it once outside of here and pass it as an argument to the method, or have a static property for the DB connection that you call a method to set.

File: src/Products/DetailedProductEntity.php [hunk 9]
```diff
@@ -0,0 +1,12 @@
+<?php
+namespace FurnitureStoreApi\Products;
+use FurnitureStoreApi\Products\SimpleProductEntity as SimpleProductEntity;
+
```
Comment (matt-ioa) [inline src/Products/DetailedProductEntity.php:4]: Bit of scope creep here: why create this if it isn't part of the story?

File: src/Categories/CategoryHydrator.php [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+        $query = $db->prepare('SELECT `categories`.`id`, `category`, COUNT(`categories`.`id`) AS `category_id` FROM `categories`
+                            JOIN `products` ON `products`.`name`= `categories`.`category`
```
Comment (matt-ioa) [inline src/Categories/CategoryHydrator.php]: Please add the category IDs to the product table and rewrite this query to join on ids instead.

File: src/Products/ProductHydrator.php [hunk 11]
```diff
@@ -0,0 +1,14 @@
+<?php
+namespace FurnitureStoreApi\Products;
+use \PDO;
+
+class ProductHydrator
+{
+    public static function getProductsByCategory($PDO, $db)
```
Comment (matt-ioa) [inline src/Products/ProductHydrator.php]: Looks like scope creep again: is this needed yet?

Also, looks like you have two parameters here: `$PDO` and `$db`. Is this what you want? And how are they different?

File: tests/ExampleTest.php [hunk 12]
```diff
@@ -0,0 +1,13 @@
+<?php
+use PHPUnit\Framework\TestCase;
+
+class ExampleTest extends TestCase {
```
Comment (matt-ioa) [inline tests/ExampleTest.php]: Delete this example test file

**Review by matt-ioa — CHANGES_REQUESTED** [review]
Quite a few changes requested, take a look

File: src/Products/ProductHydrator.php (hunk 11 above)
Comment (Henry-Tottle) [reply src/Products/ProductHydrator.php]: Yes, we realised as we did peer review but will leave it in until we get to that Story.  If we can do that?

File: src/Products/DetailedProductEntity.php (hunk 9 above)
Comment (Henry-Tottle) [reply src/Products/DetailedProductEntity.php:4]: We had this as a task to build to the spec of our architecture and got carried away creating tasks.

File: src/Services/ResponseService.php [hunk 13]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace FurnitureStoreApi\Services;
+class ResponseService
+{
+    public static function responseToJson($message, $jsonString, $responseCode)
```
Comment (matt-ioa) [inline src/Services/ResponseService.php]: Maybe a better name for this would be 'makeResponse' or something similar.

File: src/Categories/CategoryEntity.php [hunk 14]
```diff
@@ -0,0 +1,35 @@
+<?php
+namespace FurnitureStoreApi\Categories;
+use FurnitureStoreApi\Services\HeaderService;
+HeaderService::setHeaders();
+
+class CategoryEntity implements \JsonSerializable
+{
+    private int $id;
+    private string $name;
+    private int $products;
```
Comment (matt-ioa) [inline src/Categories/CategoryEntity.php]: I know that this is called 'products' in the resulting JSON, but it would be more descriptive to call it 'productCount' in this entity if you can, just to give people a better idea of what it is.

File: vendor/composer/LICENSE
Comment (matt-ioa) [inline vendor/composer/LICENSE:1]: How did the `vendor` directory make its way back in?!

File: .gitignore [hunk 15]
```diff
@@ -1,2 +1,3 @@
 .idea/
 .DS_Store
+
```
Comment (matt-ioa) [inline .gitignore]: Add that `vendor/` directory into the `.gitignore`

**Review by matt-ioa — CHANGES_REQUESTED** [review]
Just a couple more changes to make, mostly just renaming things

**Review by matt-ioa — APPROVED** [review]
🚢 it
